### DISCUSSION TITLE
SLO + Management Zone Existence Timing Bug

### DISCUSSION THREAD

**Discussion by ghost** [Dynatrace/dynatrace-configuration-as-code#333] [Archive] [answered]
**Describe the bug**
When I use monaco to create a management zone **and** an SLO which utilises that management zone as a filter, the Dynatrace API throws a 400 because monaco has not yet created the zone.

```
ERROR 	Failed to create DT object appone availability (HTTP 400)!
    Response was: {"error":{"code":400,"message":"SLO validation failed with 1 error(s):\nGiven filter is invalid."}}, responsible config: projects/baseconfig/slo/synthetic_availability.json
```

For example, my SLO filter is:

```
mzName(appone),type(HTTP_CHECK)
```

And I've also asked monaco to create the `appone` management zone. Monaco will attempt to create the SLO first:

```
INFO  Processing environment preproduction...
INFO  	Processing project projects/baseconfig...
DEBUG 		Deploying configs in this order: 
DEBUG 			1: projects/baseconfig/slo/synthetic_availability.json
DEBUG 			2: projects/baseconfig/management-zone/app.json
```

**Potential Solutions**
I believe there are two potential solutions:

1. Create the zone before the SLO or
2. Ask for DT API to be modified so that it doesn't check the existence of a management zone first and allows the SLO to be created "in good faith"

**Comment by ghost** [comment]
There is a third potential solution but it involves **not** using a management zone (so not strictly a workaround for this issue but it'll get you working again).

My scenario above was that a synthetic script would be created which is tagged as `app:whatever`. My management zone includes `HTTP_CHECK` objects with the `app:whatever` tag so instead of using the `mzName` filter, I just directly filter `HTTP_CHECK` for ones with that tag value:

```
- filter: "type(HTTP_CHECK),tag(app:whatever)"
```

**Comment by ghost** [comment]
Spoken to DT SLO team and they enforce that the filter is valid.
What that means for Monaco is that the zone must exist before the SLO is created - so I think the fix here is to ensure any zones are deployed before SLOs.

**Comment by tatomir146** [comment, marked as answer]
Hi @Dynatrace-Adam-Gardner, if you have both MZ and SLO defined in monaco, you just need to reference mzName in SLO yaml as: `zone-name: /my-project/management-zone/my-zone.name` and monaco will take care of deploying SLO after MZ is created.
